pymc/sampling/mcmc.py

```diff
@@ -914,18 +926,69 @@ def _sample_return(
     keep_warning_stat: bool,
     idata_kwargs: dict[str, Any],
     model: Model,
-) -> InferenceData | MultiTrace:
+) -> InferenceData | MultiTrace | ZarrTrace:
     """Pick/slice chains, run diagnostics and convert to the desired return type.
 
     Final step of `pm.sampler`.
     """
+    if isinstance(traces, ZarrTrace):
+        # Split warmup from posterior samples
+        traces.split_warmup_groups()
+
+        # Set sampling time
+        traces.sampling_time = t_sampling
+
+        # Compute number of actual draws per chain
+        total_draws_per_chain = traces._sampling_state.draw_idx[:]
+        n_chains = len(traces.straces)
+        desired_tune = traces.tuning_steps
+        desired_draw = len(traces.posterior.draw)
+        tuning_steps_per_chain = np.clip(total_draws_per_chain, 0, desired_tune)
+        draws_per_chain = total_draws_per_chain - tuning_steps_per_chain
+
+        total_n_tune = tuning_steps_per_chain.sum()
+        total_draws = draws_per_chain.sum()
```

Comment:
When the `ZarrTrace` groups are created, they all get a fill value and the final shape for each array. The fill value is not actually stored in the `zarr.store`, so only the actual draws take up any memory. That being said, when you open and load an unfinished run, the arrays in the trace will have the shape as if they had been completely sampled, but the unsampled draws will be set to the fill value.

Reply:
What is the default value?

Reply:
The default values are [here](https://github.com/lucianopaz/pymc/blob/zarr/pymc/backends/zarr.py#L225-L232)

Reply:
I don't like this. I think what MultiTrace is doing is more sensible. You won't get junk draws if you interrupt earlier. Here you'll get zeros and false for integers and booleans, for chains that are incomplete.

Reply:
Yes, you'll get NaNs but you'll also have the information of what was the last draw that was taken in the trace, so it will be easy to start from where you left off. Xarray isn't great for ragged arrays. What NDArray does is to cut down the chains until you shortest one you had, but we can't do that without losing the sampling state information that is needed for a re-start.

Reply:
Hmm, that's a good point. Then feel free to leave as is or issue a user warning if you think that could avoid troubles to users.